### DISCUSSION TITLE
Could we have a new circuit port creating method by self-defined pin groups?

### DISCUSSION THREAD

**Discussion by raindigoa** [ansys/pyaedt#2112] [Ideas]
Hello AE,

Here are two suggestions for edb function.

The first one is about a quite basic but unrealized method to have a new circuit port.
So far we could create the new circuit ports by the following methods while we use edb.siwave:
1. create_circuit_port_on_net
2. create_circuit_port_on_pin
3. create_circuit_port_on_components_no_pin_group

But none of the above methods allows user to create circuit port by a self-defined pin group.
It's not unusual case that we need to partition a whole bunch of pins into smaller groups to have a detailed study, so it would be nice to have such a function to do that.

The second request is about the port list in edb module, it seems there is no command that we can use to have an existed port list. If it is possible, I hope that we could have a command to list the existed ports and the corresponding positive/negative pins.

Would you please add these functions? Thank you.

**Comment by hui-zhou-a** [comment]
@raindigoa, are below methods you are looking for?
self.edbapp.core_siwave.create_current_source_on_pin_group()
self.edbapp.core_siwave.create_voltage_source_on_pin_group("sink_pos", "gnd", "vrm_voltage_source")

self.edbapp.excitations

**Reply by raindigoa** [reply]
Hi ring,
Thank you for the reply

Creating I/V source based on the assigned pin groups have already existed in the pyaedt command, but what I want to do is create a "port" on desired pin groups. It seems like there is no pyaedt command could achieve this so far.

About the second method, yes, edbapp.excitations would feedback me the existed port list as what I need, but it does not have the information of the pin groups or the any further detail except the port name, so I think it could be a to-be-added function.

Thank you again for spending time on this post.

**Comment by hui-zhou-a** [comment]
@raindigoa, new method is implemented to create circuit port on pin groups. It will take some time to be merged into main.
`self.edbapp.core_siwave.create_circuit_port_on_pin_group(
                "PG_V1P0_S0", "PinGroup_2", impedance=50, name="test_port"
            )`

When the port is created, you can tell on which net the port is placed on. 
```
port.net_name
port.reference_net_name
```
For more details about the port, the simplest way is to add info in its name. If this is not what you expect, I will need to understand your use case.

**Reply by raindigoa** [reply]
Hi ring, thank you for the effort,

but I would like to report an issue,

it seems the functions:
edbapp.core_siwave.create_voltage_source_on_pin_group()
edbapp.core_siwave.create_current_source_on_pin_group()
edbapp.core_siwave.create_circuit_port_on_pin_group()

are all failed as they are used on the "existed" pin groups.

For example, if I import an EDB file without any pin groups (edbapp.core_siwave.pin_groups = {})
and create pin group A and A_ref by assigning the pins(A1,A2 for group A, Aref1, Aref2 for group A_ref),
then I can create port by the command "create_circuit_port_on_pin_group" as what it supposes to.

However, if I open siwave or hfss3dlayout first, set the pin groups as defined above, export EDB and import it by pyaedt, then I cannot create port on the pin groups by the same command, it will still feedback us "true" just like successes. But if you check the port by edbapp.core_siwave.excitations, then you will find the port is not created.

I think this might be a bug, could you help fix it?

**Comment by hui-zhou-a** [comment]
@raindigoa pin groups created in SIWAVE GUI do have complete information. It could be the cause.
<img width="437" alt="image" src="https://user-images.githubusercontent.com/27995305/210612540-45ca858f-059c-4796-8110-4b5b9a1c8ca1.png">

I am creating an issue about it.

**Comment by hui-zhou-a** [comment]
@raindigoa fixed. https://github.com/pyansys/pyaedt/pull/2238
